lightos driver: add lightos volume driver

yogev-lb opened the pull request:
initial commit to add our driver to the cinder fork

in order to run the unittests:

Test all lightos ut:

```bash
source .tox/py38/bin/activate
stestr run -n cinder.tests.unit.volume.drivers.lightos.test_lightos_storage.LightOSStorageVolumeDriverTest
stestr run -n cinder.tests.unit.volume.drivers.lightos.test_lightos_storage.LightOSStorageVolumeDriverTest.<test_name>
```

Test specific method by name:

```bash
source .tox/py38/bin/activate
stestr run -n cinder.tests.unit.volume.drivers.lightos.test_lightos_storage.LightOSStorageVolumeDriverTest.<test_name>
```

yogev-lb committed:
lightos driver: add lightos volume driver

yogev-lb committed:
lightos: unittests - initial commit

add unittests for the lightos cinder driver.
try to mock as littel as possible by overriding the send_cmd method of
the connector.

yogev-lb committed:
drivers/lightos: dependency injection for lightos_client and connector

in order to mock these two pieces of code we must inject them from
outside.

ronen-lb commented on cinder/volume/drivers/lightos/constants.py:
```diff
@@ -0,0 +1,3 @@
+DEVICE_SCAN_ATTEMPTS_DEFAULT = 5
+LIGHTOS = "LIGHTOS"
+LIGHTOS_API_SERVICE_TIMEOUT = 30
```
newline

yogev-lb committed:
lightos/lightos.py: extract apiservers init from driver to LightOSConnection

add some validation where needed for input values from config.

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,154 @@
+        self.driver.configuration.lightos_api_address = ""
+        self.assertRaises(exception.InvalidParameterValue, lightos.LightOSConnection, self.driver.configuration)
+
+    # Yogev
```
there are some local comments as well as commented out code that should be removed

yogev-lb committed:
lightos/lightos.py: fix wait for volume delete race

wait for volume delete and wait for volume create behave differently.
NOT_FOUND means different things between create and delete.
for create it means that we didn't start yet.
for delete it means that the volume is gone and nothing to do.

yogev-lb committed:
lightos/lightos.py: create volume may return http.OK or http.CREATED

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,285 @@
+FAKE_LIGHTOS_CLUSTER_INFO: Dict[str, str] = {
+    'UUID': "926e6df8-73e1-11ec-a624-07ba3880f6cc",
+    'subsystemNQN': "nqn.2014-08.org.nvmexpress:NVMf:uuid:"\
+        "f4a89ce0-9fc2-4900-bfa3-00ad27995e7b"
```
possibly also init statistics, to enable future coverage of get_Stats

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,285 @@
+                    "src_snapshot_name": kwargs["src_snapshot_name"],
+                    "acl": kwargs["acl"],
+                    "UUID": str(uuid.uuid4()),
+                    "state": "Available",
```
not critical but real lightOS response will return the as Creating.
we want it as avaialble in DB. 
so can possibly init here to creating. 
and clone object update cloned pbject to available before adding to db.

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,285 @@
+                    "UUID": str(uuid.uuid4()),
+                    "state": "Available",
+                }
+                project = self.db["projects"].setdefault(project_name, {})
```
so create  a volume will also set/create a project in db?

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,285 @@
+                project_name = kwargs["project_name"]
+                volume_uuid = kwargs["volume_uuid"]
+                project = self.db["projects"].get(project_name, None)
+                if project:
```
if we dont find project we seem to fall down to raise runtime error.
i think close emulation of product is
 return (httpstatus.INVALID_ARGUMENT, None)

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,285 @@
+                if project:
+                    volume = next(
+                        iter(
+                            volume for volume in project["volumes"] \
```
why not use a dict under project["volumes"] with volume uuid as key?

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -0,0 +1,285 @@
+                project_name = kwargs["project_name"]
+                volume_uuid = kwargs["volume_uuid"]
+                project = self.db["projects"].get(project_name, None)
+                if project:
```
see comment above about project not found

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -57,6 +57,54 @@ def factory(protocol, root_helper, driver=None,
+
+    def __init__(self):
+        self.data = {
+            "projects"
```
to initialist it as a map shouldnt it be "projects" : {} ?

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -57,6 +57,54 @@ def factory(protocol, root_helper, driver=None,
+    def create_volume(self, volume):
+        assert volume.project_name and volume.name, "must be provided"
+        project = self.get_or_create_project(volume.project_name)
+        volumes = project.setdefault("volumes", {})
```
create volume in lightOS is not idempotent we will return error if already exist

ronen-lb commented on cinder/tests/unit/volume/drivers/lightos/test_lightos_storage.py:
```diff
@@ -57,6 +57,54 @@ def factory(protocol, root_helper, driver=None,
+        proj_vols = project.get("volumes", None)
+        if not proj_vols:
+            return
+        proj_vols.remove(volume)
```
lightos wil lreturn err if volume is not found (likely something we need to change)

ronen-lb commented on cinder/volume/drivers/lightos/lightos.py:
```diff
@@ -619,18 +619,22 @@ def _wait_for_volume_available(
-            states=states_to_wait_for,
-            vol_uuid=vol_uuid,
-            vol_name=vol_name)
+        states = ('Available', 'Deleting', 'Deleted', 'Failed', 'UNKNOWN')
```
note that lightOS will never return Deleted

not for now, but need to discuss what we want to do for states rollback and in particular migrating

ronen-lb commented on cinder/volume/drivers/lightos/lightos.py:
```diff
@@ -619,18 +619,22 @@ def _wait_for_volume_available(
+                                              vol_name=vol_name)
+            state = resp.get('state', 'UNKNOWN') if \
+                status_code == httpstatus.OK and resp else 'UNKNOWN'
+            if state in states and status_code != httpstatus.NOT_FOUND:
```
why do we break here on any err exepect for not_found?
i.e for internal error (etcd down) we want to retry.
possibly the only two cases we dont want to retry are if we fail authntication (shouldnt happen happy flow)
or invalide arg (no name and no uuid, should also not happen happy flow)

yogev-lb committed:
drivers/lightos: extend_volume quit retry if volume not found

ronen-lb commented on cinder/volume/drivers/lightos/lightos.py:
```diff
@@ -834,9 +809,24 @@ def _wait_for_snapshot_state(
-            project_name, timeout, states=(
-                'Deleted', 'Deleting', 'UNKNOWN'), vol_uuid=vol_uuid)
+        assert vol_uuid, 'LightOS volume UUID must be specified'
+        states = ('Deleted', 'Deleting', 'UNKNOWN')
```
lightOS wont return state Deleted, either one of the active volume states (Aviaalble, updating, creating, migrating, rollback), failed state or deleting
deleted volume will return unkown

ronen-lb commented on cinder/volume/drivers/lightos/lightos.py:
```diff
@@ -834,9 +809,24 @@ def _wait_for_snapshot_state(
+                return 'Deleted'
+            state = resp.get('state', 'UNKNOWN') if \
+                status_code == httpstatus.OK and resp else 'UNKNOWN'
+            if state in states:
```
thnk this would be clearr if you just check state== deleting (deleted and unkown should never be returned)

ronen-lb commented on cinder/volume/drivers/lightos/lightos.py:
```diff
@@ -738,7 +738,7 @@ def _create_volume(self, volume, src_snapshot_lightos_name):
 
         # when we get here, we definitely tried to create it in the past
-        if status_code == httpstatus.OK:
+        if status_code in (httpstatus.OK, httpstatus.CREATED):
```
this is for UT? lighttOS can return status created?

yuval-lb committed:
opendev code

yogev-lb committed:
lightos/lightos.py: fix wait for snapshots delete race

Wait for snapshot delete and wait for snapshot create behave
differently.
NOT_FOUND means different things between create and delete.
for create it means that we didn't start yet.
for delete it means that the snapshot is gone and nothing to do.

tomerg-lb reviewed (commented):
@yogev-lb , is this still relevant? If not, please close
